### PR TITLE
fix: Report errors to host during wallet reconstruction flows

### PR DESCRIPTION
**Pull request by kartikCypherock**
Implements https://app.clickup.com/t/9002019994/PRF-4892

Requires https://github.com/Cypherock/cypherock-common/pull/34

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4892 Wallet state syncing](https://app.clickup.com/t/9002019994/PRF-4892)

File: src/wallet/wallet_list.c [hunk 1]
```diff
@@ -139,22 +139,29 @@ uint8_t get_valid_wallet_meta_data_list(wallet_list_t *list) {
...
+                           Wallet *wallet,
+                           rejection_cb *reject_cb) {
   if ((NULL == wallet_id) || (NULL == wallet)) {
     return false;
   }
 
   for (uint8_t wallet_idx = 0; wallet_idx < MAX_WALLETS_ALLOWED; wallet_idx++) {
     wallet_state state = INVALID_WALLET;
-    if (true != wallet_is_filled(wallet_idx, &state)) {
+    if (wallet_is_filled(wallet_idx, &state)) {
```
Comment (amanCypherock) [inline src/wallet/wallet_list.c:151]: Recheck condition for continue.

File: src/wallet/wallet_list.c [hunk 2]
```diff
@@ -139,22 +139,29 @@ uint8_t get_valid_wallet_meta_data_list(wallet_list_t *list) {
...
 
-    if (VALID_WALLET == state) {
-      if (0 == memcmp(wallet_id, get_wallet_id(wallet_idx), WALLET_ID_SIZE)) {
-        if (true == is_wallet_locked(wallet_idx)) {
-          return false;
+    if (0 == memcmp(wallet_id, get_wallet_id(wallet_idx), WALLET_ID_SIZE)) {
+      // Check if wallet is in usable state or not
+      if (INVALID_WALLET == state || is_wallet_locked(wallet_idx) ||
+          is_wallet_partial(wallet_idx) || is_wallet_unverified(wallet_idx) ||
+          is_wallet_share_not_present(wallet_idx)) {
```
Comment (amanCypherock) [inline src/wallet/wallet_list.c:159]: Why not simply check 
`(VALID_WALLET != state) || is_wallet_locked(wallet_idx) || is_wallet_partial(wallet_idx)`

File: src/wallet/wallet_list.c [hunk 3]
```diff
@@ -163,22 +170,49 @@ bool get_wallet_data_by_id(const uint8_t *wallet_id, Wallet *wallet) {
...
+          reject_cb(ERROR_COMMON_ERROR_WALLET_PARTIAL_STATE_TAG, 0);
+        }
+        return false;
+      } else {
+        if (wallet_name) {
+          memcpy(wallet_name, get_wallet_name(wallet_idx), NAME_SIZE);
+        }
+        return true;
+      }
+    }
```
Comment (amanCypherock) [inline src/wallet/wallet_list.c:209]: This looks like logic duplication, consider using an implementation similar to previous one where we call `get_wallet_data_by_id` to fetch name.

File: src/wallet/reconstruct_wallet_flow.c [hunk 4]
```diff
@@ -333,20 +333,33 @@ bool reconstruct_seed_flow(const uint8_t *wallet_id, uint8_t *seed_out) {
...
+  if (COMPLETED == flow) {
     get_seed_from_secret(secret, seed_out);
     result = true;
+  } else if (reject_cb && EARLY_EXIT == flow) {
+    // Inform the host of any rejection
+    reject_cb(ERROR_COMMON_ERROR_USER_REJECTION_TAG,
+              ERROR_USER_REJECTION_CONFIRMATION);
+  } else if (reject_cb && COMPLETED_WITH_ERRORS == flow) {
+    // Inform the host of any card error
+    reject_cb(ERROR_COMMON_ERROR_CARD_ERROR_TAG, 0);
```
Comment (amanCypherock) [inline src/wallet/reconstruct_wallet_flow.c:353]: Let's use the enum `WALLET_NOT_FOUND_UNKNOWN` or `WALLET_PARTIAL_STATE_UNKNOWN` for reference.

File: src/wallet/wallet_list.c (hunk 2 above)
Comment (kartikCypherock) [reply src/wallet/wallet_list.c:159]: Thanks for this suggestion

File: src/wallet/wallet_list.c (hunk 3 above)
Comment (kartikCypherock) [reply src/wallet/wallet_list.c:209]: Got it, updated

File: src/wallet/reconstruct_wallet_flow.c (hunk 4 above)
Comment (kartikCypherock) [reply src/wallet/reconstruct_wallet_flow.c:353]: Ok, updated

**Review by amanCypherock — APPROVED** [review]

**Pull request merged**